developer/src/kmcmplib/src/Compiler.cpp

```diff
@@ -0,0 +1,3868 @@
+  kmcmp::nErrors = 0;
+
+  fp_in = fopen((const  KMX_CHAR*)pszInfile, "rb");
+
+  if (fp_in == NULL) SetError(CERR_InfileNotExist);
+
+
+  // Transfer the file to a memory stream for processing UTF-8 or ANSI to UTF-16?
+  // What about really large files?  Transfer to a temp file...
+  fseek(fp_in, 0, SEEK_END);
+  len = ftell(fp_in);
+  fseek(fp_in, 0, SEEK_SET);
+  if (!fread(str, 1, 3, fp_in))
+  {
+    fclose(fp_in);
+    return CERR_CannotReadInfile;
+  }
+
+  fseek(fp_in, 0, SEEK_SET);
+  if (str[0] == UTF8Sig[0] && str[1] == UTF8Sig[1] && str[2] == UTF8Sig[2])
+    fp_in = UTF16TempFromUTF8(fp_in, TRUE);
+  else if (str[0] == UTF16Sig[0] && str[1] == UTF16Sig[1])
+    fseek(fp_in, 2, SEEK_SET);
+  else
+    fp_in = UTF16TempFromUTF8(fp_in, FALSE);
+  if (fp_in == NULL)
+  {
+    return CERR_CannotCreateTempfile;
+  }
+
+  fp_out = fopen((const  KMX_CHAR*)pszOutfile, "wb");
+
+  if (fp_out == NULL) SetError(CERR_CannotCreateOutfile);
+
+
+  KMX_DWORD msg;
+  FILE_KEYBOARD fk;
+  CodeConstants = new kmcmp::NamedCodeConstants;
+  err = CompileKeyboardHandle(fp_in, &fk);
+  if (err)
+  {
+    if ((msg = WriteCompiledKeyboard(&fk, fp_out)) != CERR_None)
+      AddCompileError(msg);
+  }
+  else
+    AddCompileError(CERR_InvalidValue);
+
+  fclose(fp_in);
+  fclose(fp_out);
+
+  delete CodeConstants;
+
+  if (kmcmp::nErrors > 0)
+  {
+    remove(pszOutfile);
+    return FALSE;
+  }
+
+  return err;
+}
+
+
+
+extern "C" BOOL __declspec(dllexport)  kmcmp_CompileKeyboardFileToBuffer(PKMX_STR pszInfile, PFILE_KEYBOARD pfkBuffer, KMX_BOOL ACompilerWarningsAsErrors, KMX_BOOL AWarnDeprecatedCode, CompilerMessageProc pMsgProc, int Target)   // I4865   // I4866
+{
+  printf("°°-> changed to CompileKeyboardFileToBuffer() of kmcmplib \n");
+  FILE* fp_in = NULL;
+  KMX_BOOL err;
+  KMX_DWORD len;
+  KMX_DWORD len2;
+  KMX_CHAR str[260];
+
+  kmcmp::FSaveDebug = TRUE;   // I3681
+  kmcmp::FCompilerWarningsAsErrors = ACompilerWarningsAsErrors;   // I4865
+  kmcmp::FWarnDeprecatedCode = AWarnDeprecatedCode;   // I4866
+
+  kmcmp::CompileTarget = Target;
+
+  if (!pMsgProc || !pszInfile || !pfkBuffer) SetError(CERR_BadCallParams);
+
+  PKMX_STR p;
+
+#if defined(_WIN32) || defined(_WIN64)
+  if (p = strrchr(pszInfile, '\\'))
+#else
+  if (p = strrchr(pszInfile, '/'))
+#endif
+  {
+    strncpy_s(kmcmp::CompileDir, _countof(kmcmp::CompileDir), pszInfile, (INT_PTR)(p - pszInfile + 1));  // I3481
+    kmcmp::CompileDir[(INT_PTR)(p - pszInfile + 1)] = 0;
+  }
+  else
+    kmcmp::CompileDir[0] = 0;
+
+  msgproc = pMsgProc;
+  kmcmp::currentLine = 0;
+  kmcmp::nErrors = 0;
+
+  fp_in = fopen(pszInfile,"rb");
+
+  if (fp_in == NULL) SetError(CERR_InfileNotExist);
+
+  // Transfer the file to a memory stream for processing UTF-8 or ANSI to UTF-16?
+  // What about really large files?  Transfer to a temp file...
+
+  KMX_DWORD sz = ftell(fp_in);
+  len2 = fread(str,1,sz,fp_in);
+  if( !len2)
+  {
+    fclose(fp_in);
+    return CERR_CannotReadInfile;
+  }
```

Comment:
This was causing all web keyboards to fail to compile.

```suggestion
  if(!fread(str,1,3,fp_in)) {
    fclose(fp_in);
    return CERR_CannotReadInfile;
  }
```

Reply:
OK